drivers/ethernet/dwc_xgmac/eth_dwc_xgmac.c

```diff
@@ -0,0 +1,1992 @@
+#ifdef CONFIG_NET_STATISTICS_ETHERNET
+	/*
+	 * Ethernet statistics captured by XGMAC driver
+	 */
+	struct net_stats_eth stats;
+#endif /* CONFIG_NET_STATISTICS_ETHERNET */
+#ifdef CONFIG_ETH_DWC_XGMAC_POLLING_MODE
+	/*
+	 * timer for interrupt polling
+	 */
+	struct k_timer isr_polling_timer;
+#endif /* CONFIG_ETH_DWC_XGMAC_POLLING_MODE */
+#ifdef CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE
+	/*
+	 * work que item for processing TX interrupt bottom half
+	 */
+	struct k_work isr_work;
+#endif /*CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE*/
+	struct k_mutex dev_cfg_lock;
+};
+
+/**
+ * @brief Constant device configuration data structure.
+ *
+ * This struct contains all device configuration data for a XGMAC
+ * controller instance which is constant. The data herein is
+ * either acquired from the generated header file based on the
+ * data from Kconfig, or from header file based on the device tree
+ * data. Some of the data contained, in particular data relating
+ * to clock sources, is specific to the platform, which contain the XGMAC.
+ */
+struct eth_dwc_xgmac_config {
+	DEVICE_MMIO_ROM;
+	/*
+	 * Use a random MAC address generated when the driver is initialized
+	 */
+	bool random_mac_address;
+	/*
+	 * Number of TX queues configured
+	 */
+	uint8_t num_tx_Qs;
+	/*
+	 * Number of RX queues configured
+	 */
+	uint8_t num_rx_Qs;
+	/*
+	 * Number of DMA channels configured
+	 */
+	uint8_t num_dma_chnl;
+	/*
+	 * Number of traffic classes configured
+	 */
+	uint8_t num_TCs;
+	/*
+	 * Maximum transfer unit length configured
+	 */
+	uint16_t mtu;
+	/*
+	 * Transmit FIFO size
+	 */
+	uint32_t tx_fifo_size;
+	/*
+	 * Receive FIFO size
+	 */
+	uint32_t rx_fifo_size;
+	/*
+	 * XGMAC DMA configuration
+	 */
+	struct xgmac_dma_cfg dma_cfg;
+	/*
+	 * XGMAC DMA channels configuration
+	 */
+	struct xgmac_dma_chnl_config dma_chnl_cfg;
+	/*
+	 * XGMAC MTL configuration
+	 */
+	struct xgmac_mtl_config mtl_cfg;
+	/*
+	 * XGMAC core configuration
+	 */
+	struct xgmac_mac_config mac_cfg;
+	/*
+	 * Global pointer to traffic classes and queues configuration
+	 */
+	struct xgmac_tcq_config *tcq_config;
+	/*
+	 * Ethernet PHY device pointer
+	 */
+	const struct device *phy_dev;
+	/*
+	 * Interrupts configuration function pointer
+	 */
+	eth_config_irq_t irq_config_fn;
+	/*
+	 * Interrupts enable function pointer
+	 */
+	eth_enable_irq_t irq_enable_fn;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
+	/*
+	 * XGMAC peripheral reset signal
+	 */
+	const struct reset_dt_spec reset;
+#endif
+};
+
+static inline mem_addr_t get_reg_base_addr(const struct device *dev)
+{
+	return (mem_addr_t)DEVICE_MMIO_GET(dev);
+}
+
+static inline int dwxgmac_software_reset(mem_addr_t ioaddr)
+{
+	uint32_t delay_us = 1000;
+	bool ret;
+	/* software reset */
+	mem_addr_t reg_addr = (mem_addr_t)(ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_MODE_OFST);
+
+	sys_write32(DMA_MODE_SWR_SET(1u), reg_addr);
+
+	ret = WAIT_FOR(!(sys_read32(reg_addr) & DMA_MODE_SWR_SET_MSK), delay_us, k_msleep(1));
```

Comment:
same suggestion as the mdio driver about configurability with kconfig

Reply:
added a config item for timeout